text/0580-destroyables.md

```diff
@@ -0,0 +1,243 @@
+   attempting to patch `willDestroy` in this way.
+
+4. Finally, since this isn't a standard, it's difficult to add _layers_ of new
+   destroyable values that can interoperate with one another. For instance,
+   there is no way for `ember-concurrency` to know how to destroy tasks on
+   non-framework classes that users may have added themselves.
+
+This RFC proposes a streamlined API that disconnects the exact implementation
+from any interface, allows for multiple destructors per-destroyable, and
+maximizes interoperability in general.
+
+## Detailed design
+
+The API consists of 6 main functions, imported from `@ember/application`:
+
+```ts
+declare function associateDestroyableChild(parent: object, child: object): void;
+declare function registerDestructor(destroyable: object, destructor: () => void): void;
+declare function unregisterDestructor(destroyable: object, destructor: () => void): void;
+
+declare function destroy(destroyable: object): void;
+declare function isDestroying(destroyable: object): boolean;
+declare function isDestroyed(destroyable: object): boolean;
+```
+
+In addition, there is a debug-only mode function used for testing:
+
+```ts
+declare function assertDestroyablesDestroyed(): void;
+```
+
+#### `associateDestroyableChild`
+
+This function is used to associate a destroyable object with a parent. When the
+parent is destroyed, all registered children will also be destroyed.
+
+- Attempting to associate a parent or child that has already been destroyed will
+  throw an error.
+- Attempting to associate a child with a parent that is not a destroyable should
+  throw an error.
+
+##### Multiple Inheritance
+
+Attempting to associate a child to multiple parents should currently throw an
+error. This could be changed in the future, but for the time being multiple
+inheritance of destructors is tricky and not scoped in. Instead, users can add
+destructors to accomplish this goal:
+
+```js
+let parent1 = {}, parent2 = {}, child = {};
+
+registerDestructor(parent1, () => destroy(child));
+registerDestructor(parent2, () => destroy(child));
+```
+
+The exact timing semantics here will be a bit different, but for most use cases
+this should be fine. If we find that it would be useful to have multiple
+inheritance baked in in the future, it can be added in a followup RFC.
+
+#### `registerDestructor`
+
+Receives a destroyable and a destructor function, and associates the function
+with it. When the destroyable is destroyed, or when its parent is destroyed, the
+destructor function will be called. Multiple destructors can be associated with
+a given destroyable, and they can be associated over time, allowing libraries
+like `ember-lifeline` to dynamically add destructors as needed.
+
+- Registering a destructor on a destroyed object should throw an error.
+- Attempting to register the same destructor multiple times should throw an
+  error.
+- Attempting to register a destructor on a non-destroyable should throw an
```

Comment:
I'd like a more concrete definition of "destroyable" (similar to my other comment above)